### PR TITLE
feat(storybook): storybook 7

### LINKED ISSUE #2266: [feat] Upgrade Storybook (version 7+)
**Is your feature request related to a problem? Please describe.**
We are a major version behind with storybook. This might be a dependency we want to stay more up to date on because storybook has several dependencies related to it that trigger dependency alerts and increase the noise we have to sort through.

We are already seeing some [alerts ](https://github.com/trussworks/react-uswds/security/dependabot/72)for outdated deps attached the version of storybook we are on. In the example case, dependabot was unable to update because of conflicting versions between storybook deps.

If we upgrade all of storybook it should resolve this issue. The (`terser-webpack-plugin` and related terser-x deps) are a full major version higher on recent versions of storybook.  lso of note - (storybook [has specific guidelines for how to upgrade]( https://storybook.js.org/docs/react/configure/upgrading) in sync). 

**Describe the solution you'd like**
Go up to Storybook v 7.x to stay up to date. This would necessarily include updating all related storybook dependencies.

### PR DESCRIPTION
**Pull request by jpandersen87**
# Summary

Upgrades project's Storybook to version 7.1. Storybook's auto-migrations were ran which affected configuration and stories (function to object-based stories). 

In addition:
- internal-only stories had their filenames changed (*.stories.tsx -> *.stories.internal.tsx) to prevent Storybook from attempting to load them with the matching glob pattern (which caused warnings as Storybook would fail to load the commented-out default export).
- The USWDS svg referenced from USWDS' github hosting has been copied to a `.storybook/public` folder in order to reference locally instead (in the documentation story, and as the branding logo in `USWDS.js`)
- adds manual version resolution for jackspeak to fix yarn v1 bug with storybook 7.1 (see https://github.com/storybookjs/storybook/issues/22431#issuecomment-1630086092 )
- updates typescript to 5.1.6 (fixes transient dependency minimum typescript issues)
  - some new type errors in project were corrected that appeared from update
- updates @typescript-eslint dependencies to 6.2.0
- updates jest to 29.6.1
- updates ts-jest to 29.1.1
- adds jest-environment-jsdom

## Related Issues or PRs

Closes #2266

## How To Test

Run storybook locally via `yarn run storybook`

### CODE REVIEW

File: .storybook/main.ts [hunk 1]
```diff
@@ -71,20 +73,17 @@ const webpackConfig = (config) => {
...
   stories: ['../src/**/*.stories.@(ts|tsx)'],
   addons: ['@storybook/addon-essentials', '@storybook/addon-a11y'],
-  typescript: {
-    check: false,
-    checkOptions: {},
-    reactDocgen: 'react-docgen-typescript',
-    reactDocgenTypescriptOptions: {
-      tsconfigPath: path.resolve(__dirname, '../tsconfig.json'),
-    },
-  },
```
Comment (gidjin) [inline .storybook/main.ts:86]: Question: Why was this section removed?

File: .storybook/main.ts [hunk 2]
```diff
@@ -71,20 +73,17 @@ const webpackConfig = (config) => {
...
   webpackFinal: async (config) => {
     return webpackConfig(config)
   },
-}
+  framework: {
+    name: '@storybook/react-webpack5',
+    options: { strictMode: false },
+  },
+  docs: {
+    autodocs: true,
```
Comment (gidjin) [inline .storybook/main.ts:86]: Suggestion: I don't think we want docs generated for every component, could you update this to be `autodocs: 'tag',` instead.

```suggestion
    autodocs: 'tag',
```

File: package.json [hunk 3]
```diff
@@ -110,6 +109,7 @@
     "sass-resources-loader": "^2.0.1",
     "source-map-loader": "^4.0.1",
     "standard-version": "^9.0.0",
+    "storybook": "^7.0.23",
```
Comment (gidjin) [inline package.json:112]: Suggestion(nit-pick): Since there are other requested changes would you mind bumping this to the latest storybook 7.1.0?

**Review by gidjin — CHANGES_REQUESTED** [review]
Left some comments. Marking request changes for the questions in `.storybook/main.js`

File: .storybook/main.ts (hunk 1 above)
Comment (jpandersen87) [reply .storybook/main.ts]: An incorrect assumption that storybook default settings covered this, but I failed to notice the disabling of the check property. This section has been returned.

**Review by gidjin — APPROVED** [review]
This looks good to me, I'd like to get a Happo run and one of our designers to approve before merging.

**Comment by jpandersen87** [comment]
Made some updates due to what appeared to be issues with react-uswds' typescript version (bumped to latest version).

**Comment by gidjin** [comment]
There are some changes needed in the build files. I opened up #2530 to make those changes. Figure once that is merged in we can rebuild this. Or if you want to merge those changes into your PR, we can close the other.

**Comment by jpandersen87** [comment]
> There are some changes needed in the build files. I opened up #2530 to make those changes. Figure once that is merged in we can rebuild this. Or if you want to merge those changes into your PR, we can close the other.

I've copied those changes into this PR.

**Comment by jpandersen87** [comment]
The automated github workflows from my forked copy make it to the happo step (before obviously failing due to missing secrets), so fingers crossed that the current version is sound. Apologies on the back and forth.

**Pull request closed without merging**